core/invoker/src/main/scala/whisk/core/container/ContainerUtils.scala

```diff
@@ -164,14 +164,19 @@ trait ContainerUtils {
             val buffer = java.nio.ByteBuffer.allocate(remain)
             while (remain > 0) {
                 val read = channel.read(buffer)
-                if (read > 0)
-                    remain = read - read.toInt
+                if (read >= 0) {
+                    remain -= read
+                } else {
+                    // read < 0 means premature end of file => terminate loop
+                    remain = 0
+                }
```

Comment:
We are only talking about edge cases here.

Current line 168 is wrong and would set `remain`to zero whenever `channel.read()` was able to read something - in other words: almost always. This bug actually delegates all problems with reading to the caller of `getDockerLogContent()` eating up the retry count.

Two different cases are possible:

1. It _could_ happen that `channel.read()` is not able to fill the whole buffer in the first attempt, i.e. `0 < read < remain`. The bug in line 168 causes the loop to terminate. The calling function cannot find the sentinel marker, will drop the read buffer and will retry - invoking `getDockerLogContent()` again. This condition may arise if the log file is large and a lot of I/O is going on.

2. It _could_ happen that Docker truncates the log files after the its size has been determined by the caller of `getDockerLogContent()`. In case of full truncation, the loop is repeated until `channel.read()` returns a positive value - potentially blocking forever.

Both cases are expected to rarely happen in the wild - but if they happen, they can degrade performance considerably. And the fix is easy.

My recent commit `1e1494b847f4e0548d66db9d2108b3b514b5e082` adds unit tests.